### PR TITLE
Laboratory Work #1, Anastasia Buntova - 20FPL1

### PR DESCRIPTION
**Pull request by DangerousCrocodile**

### CODE REVIEW

File: pipeline.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 """
 Pipeline for text processing implementation
 """
+import requests
```
Comment (demid5111) [inline pipeline.py:4]: you should include it in the requirements file, better with a version

**Review by demid5111 — CHANGES_REQUESTED** [review]

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+requests>=2.27.1
```
Comment (demid5111) [inline requirements.txt:1]: 1. alphabetic order
2. no testing dependencies in requirements.txt. Instead, they should be in requirements_qa.txt
3. better to freeze exact versions (not mandatory)

File: scrapper.py [hunk 3]
```diff
@@ -2,6 +2,14 @@
 Scrapper implementation
 """
 
+import json
```
Comment (demid5111) [inline scrapper.py:5]: all imports should be grouped into:

1. standard Python imports
2. 3rd party imports
3. project imports

all imports should be in alphabetic order

File: scrapper.py [hunk 4]
```diff
@@ -2,6 +2,14 @@
...
+import re
+import os
+import time
+import requests
+from bs4 import BeautifulSoup
+# from constants import CRAWLER_CONFIG_PATH, ASSETS_PATH
```
Comment (demid5111) [inline scrapper.py:11]: we do not accept commented code. remove or uncomment

File: scrapper.py [hunk 5]
```diff
@@ -35,7 +45,23 @@ def find_articles(self):
         """
         Finds articles
         """
-        pass
+        headers = {
```
Comment (demid5111) [inline scrapper.py:48]: is there a better way to handle those configuration items? especially given they are used several times

File: scrapper.py [hunk 6]
```diff
@@ -35,7 +45,23 @@ def find_articles(self):
...
+            'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 13_2_3 like Mac OS X) '
+            'AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.3 Mobile/15E148 Safari/604.1',
+            'Accept-Encoding': 'gzip, deflate, br',
+            'Accept-Language': 'ru,en;q=0.9,de;q=0.8'
+        }
+
+        for seed_url in self.seed_urls:
+            response = requests.get(url=seed_url, headers=headers)
+            time.sleep(3)
```
Comment (demid5111) [inline scrapper.py:57]: 🏆

File: scrapper.py [hunk 7]
```diff
@@ -35,7 +45,23 @@ def find_articles(self):
...
+            if not response.ok:
+                print("Something went wrong...")
```
Comment (demid5111) [inline scrapper.py:59]: it is a weak check. probably we do not want to process such an URL then? what do you think?

File: scrapper.py [hunk 8]
```diff
@@ -35,7 +45,23 @@ def find_articles(self):
...
+            soup = BeautifulSoup(response.text, "lxml")
```
Comment (demid5111) [inline scrapper.py:60]: haven't seen that in requirements

File: scrapper.py [hunk 9]
```diff
@@ -35,7 +45,23 @@ def find_articles(self):
...
+            main_block_bs = soup.find('div', {'class': 'cat-children'})
```
Comment (demid5111) [inline scrapper.py:61]: move extraction logic to the 

```python
 def _extract_url(self, article_bs):
        pass
```

File: scrapper.py [hunk 10]
```diff
@@ -35,7 +45,23 @@ def find_articles(self):
...
+            for link in main_block_bs.find_all("a"):
+                self.urls.append('https://l.jvolsu.com/' + link.get('href'))
```
Comment (demid5111) [inline scrapper.py:63]: this constant should be also moved to some configurable place

File: scrapper.py [hunk 11]
```diff
@@ -48,14 +74,33 @@ def prepare_environment(base_path):
     """
     Creates ASSETS_PATH folder if not created and removes existing folder
     """
-    pass
+    try:
+        os.rmdir(base_path)
```
Comment (demid5111) [inline scrapper.py:78]: will work for empty dictionaries only. read the docs: https://docs.python.org/3/library/os.html#os.rmdir

File: scrapper.py [hunk 12]
```diff
@@ -48,14 +74,33 @@ def prepare_environment(base_path):
...
+    except FileNotFoundError:
+        os.mkdir(base_path)
```
Comment (demid5111) [inline scrapper.py:80]: might not work as well

File: scrapper.py [hunk 13]
```diff
@@ -48,14 +74,33 @@ def prepare_environment(base_path):
...
 
 
 def validate_config(crawler_path):
     """
     Validates given config
     """
-    pass
+    with open(crawler_path, "r", encoding="utf-8") as file:
+        config = json.load(file)
+    seed_urls = config["seed_urls"]
```
Comment (demid5111) [inline scrapper.py:89]: could it be that those keys are not in the file?

File: scrapper.py [hunk 14]
```diff
@@ -48,14 +74,33 @@ def prepare_environment(base_path):
...
+    number_of_articles = config["total_articles_to_find_and_parse"]
+    if not seed_urls or not isinstance(seed_urls, list):
+        raise IncorrectURLError
+    for seed_url in seed_urls:
+        if not isinstance(seed_url, str):
+            raise IncorrectURLError
+        norm_url = re.match(r"https?://", seed_url)
+        if not norm_url:
+            raise IncorrectURLError
```
Comment (demid5111) [inline scrapper.py:98]: would you accept such links? https://github.com/fipl-hse/2021-2-level-ctlr/pull/24/files

should you?

**Review by demid5111 — CHANGES_REQUESTED** [review]
continue to HTMLParser implementation

File: pipeline.py [hunk 15]
```diff
@@ -1,6 +1,7 @@
 """
```
Comment (demid5111) [inline pipeline.py:1]: discard changes in this file(!)

**Comment by demid5111** [comment]
Almost no progress since last time! hurry up, you are at risk of missing the first deadline!

**Comment by demid5111** [comment]
you have missed the deadline. first, make sure you finish scrapper with minimal penalty and start working on text processing pipeline, deadline will approach shortly

File: scrapper.py [hunk 16]
```diff
@@ -44,20 +76,120 @@ def get_search_urls(self):
...
+        title_and_author = page.find('h2')
+        title_and_author = title_and_author.text.replace(' "', "").replace('" ', "")
+
+        title = ""
+        for letter in title_and_author[::-1]:
+            if letter == ".":
+                if title_and_author[title_and_author.index(letter) - 1].isupper():
+                    break
+            title += letter
+        self.article.title = title[::-1][1:-2]
```
Comment (demid5111) [inline scrapper.py:122]: I do not understand the logic of taking the title. Please explain and simplify implementation

File: scrapper_config.json [hunk 17]
```diff
@@ -1,4 +1,15 @@
 {
-    "seed_urls": [],
-    "total_articles_to_find_and_parse": 0
+    "seed_urls": ["https://l.jvolsu.com/index.php/ru/archive-ru/725-science-journal-of-volsu-linguistics-2021-vol-20-no-6",
+    "https://l.jvolsu.com/index.php/ru/archive-ru/714-science-journal-of-volsu-linguistics-2021-vol-20-no-5",
+        "https://l.jvolsu.com/index.php/ru/archive-ru/701-science-journal-of-volsu-linguistics-2021-vol-20-no-4",
```
Comment (demid5111) [inline scrapper_config.json:4]: make formatting great again

**Review by demid5111 — CHANGES_REQUESTED** [review]
apply several comments and we will talk tomorrow

File: scrapper.py [hunk 18]
```diff
@@ -44,20 +76,120 @@ def get_search_urls(self):
...
+
+        author = title_and_author.replace(title[::-1], "")
+        if "," in author:
+            self.article.author = author.split(",")[0][4:]
```
[the remark this answers is not in the document]
Comment (DangerousCrocodile) [reply scrapper.py:126]: Well if I remove this slice there will be smth like "\t\s\s" in the beginning of the string

File: scrapper.py (hunk 16 above)
Comment (DangerousCrocodile) [reply scrapper.py:122]: Now I changed it but the logic is that there's the title and the author (or authors) in the same string and I need to divide them so I search for the dot which is followed with capitalized letter (in the reversed string) so this dot will be the separator (there are some articles where a dot is placed in its title so the condition that the next letter must be capitalized is important)

Comment (DangerousCrocodile) [reply scrapper.py:122]: like in "Новак М.О. Названия музыкальных инструментов в древнерусских паримейниках XII–XIV вв. (на материале исторического корпуса «Манускрипт»)"

File: scrapper_config.json [hunk 19]
```diff
@@ -1,4 +1,15 @@
...
+    "https://l.jvolsu.com/index.php/ru/archive-ru/688-science-journal-of-volsu-linguistics-2021-vol-20-no-3",
+    "https://l.jvolsu.com/index.php/ru/archive-ru/679-science-journal-of-volsu-linguistics-2021-vol-20-no-2",
+    "https://l.jvolsu.com/index.php/ru/archive-ru/668-science-journal-of-volsu-linguistics-2021-vol-20-no-1",
+    "https://l.jvolsu.com/index.php/ru/archive-ru/655-science-journal-of-volsu-linguistics-2020-vol-19-no-6",
+    "https://l.jvolsu.com/index.php/ru/archive-ru/642-science-journal-of-volsu-linguistics-2020-vol-19-no-5",
+    "https://l.jvolsu.com/index.php/ru/archive-ru/630-science-journal-of-volsu-linguistics-2020-vol-19-no-4",
+    "https://l.jvolsu.com/index.php/ru/archive-ru/619-science-journal-of-volsu-linguistics-2020-vol-19-no-3",
+    "https://l.jvolsu.com/index.php/ru/archive-ru/608-science-journal-of-volsu-linguistics-2020-vol-19-no-2",
+    "https://l.jvolsu.com/index.php/ru/archive-ru/595-science-journal-of-volsu-linguistics-2020-vol-19-no-1"],
+    "total_articles_to_find_and_parse": 17
```
[the remark this answers is not in the document]
Comment (DangerousCrocodile) [reply scrapper_config.json]: Yes, I can collect it on my computer

File: scrapper.py [hunk 20]
```diff
@@ -44,20 +77,120 @@ def get_search_urls(self):
...
 
+class HTMLParser:
+    def __init__(self, article_url, article_id):
+        self.article_url = article_url
+        self.article_id = article_id
+        self.article = Article(self.article_url, self.article_id)
+
+    def parse(self):
+        response = requests.get(self.article_url, headers=HEADERS)
+        time.sleep(random.randint(1, 3))
```
Comment (demid5111) [inline scrapper.py:88]: this should be implemented in a logic outside the parse method - somewhere where you call it

**Review by demid5111 — CHANGES_REQUESTED** [review]
address the comments!

**Comment by demid5111** [comment]
no progress with a pipeline, hurry up!

**Review by demid5111 — CHANGES_REQUESTED** [review]
where is pipeline implementation? it is time to start

**Review by demid5111 — CHANGES_REQUESTED** [review]
where is pipeline implementation? are you going to show it? time is approaching

File: pipeline.py [hunk 21]
```diff
@@ -51,19 +63,26 @@ class CorpusManager:
     """
 
     def __init__(self, path_to_raw_txt_data: str):
-        pass
+        self._storage = {}
+        self.path = Path(path_to_raw_txt_data)
```
Comment (demid5111) [inline pipeline.py]: .

File: pipeline.py [hunk 22]
```diff
@@ -51,19 +63,26 @@ class CorpusManager:
...
 
     def _scan_dataset(self):
         """
         Register each dataset entry
         """
-        pass
+        files = self.path.glob("*")
```
Comment (demid5111) [inline pipeline.py]: .intermediate variable

File: pipeline.py [hunk 23]
```diff
@@ -51,19 +63,26 @@ class CorpusManager:
...
+        pattern = re.compile(r'[0-9]+')
+        for texts in files:
+            if "_raw.txt" in texts.name:
```
Comment (demid5111) [inline pipeline.py]: make it shallow

File: pipeline.py [hunk 24]
```diff
@@ -51,19 +63,26 @@ class CorpusManager:
...
+                article_id = int(pattern.search(texts.name).group(0))
```
Comment (demid5111) [inline pipeline.py]: number is not guaranteed

File: pipeline.py [hunk 25]
```diff
@@ -51,19 +63,26 @@ class CorpusManager:
...
+        self._storage = {}
+        self.path = Path(path_to_raw_txt_data)
 
     def _scan_dataset(self):
         """
         Register each dataset entry
         """
-        pass
+        files = self.path.glob("*")
+        pattern = re.compile(r'[0-9]+')
```
Comment (demid5111) [inline pipeline.py]: using shorhand?

File: pipeline.py [hunk 26]
```diff
@@ -51,19 +63,26 @@ class CorpusManager:
...
+            if "_raw.txt" in texts.name:
+                article_id = int(pattern.search(texts.name).group(0))
+                self._storage[article_id] = Article(url=None, article_id=article_id)
 
     def get_articles(self):
         """
         Returns storage params
         """
-        pass
+        self._scan_dataset()
```
Comment (demid5111) [inline pipeline.py]: scanning should be made somewhere else - read docs

File: pipeline.py [hunk 27]
```diff
@@ -72,31 +91,100 @@ class TextProcessingPipeline:
...
     def __init__(self, corpus_manager: CorpusManager):
-        pass
+        self.corpus_manager = corpus_manager
 
     def run(self):
         """
         Runs pipeline process scenario
         """
-        pass
+        articles = self.corpus_manager.get_articles()
```
Comment (demid5111) [inline pipeline.py]: .

File: pipeline.py [hunk 28]
```diff
@@ -72,31 +91,100 @@ class TextProcessingPipeline:
...
+
+        for article in articles.values():
+            article_text = article.get_raw_text()
```
Comment (demid5111) [inline pipeline.py:101]: .

File: pipeline.py [hunk 29]
```diff
@@ -72,31 +91,100 @@ class TextProcessingPipeline:
...
         """
-        pass
+        normal_text = raw_text.replace("\n", " ")
+
+        analyzed_text = Mystem().analyze(normal_text)
+        morphs = pymorphy2.MorphAnalyzer()
+
+        tokens = []
+        for token in analyzed_text:
+            if not token.get("analysis"):
```
Comment (demid5111) [inline pipeline.py]: order of checks

File: pipeline.py [hunk 30]
```diff
@@ -72,31 +91,100 @@ class TextProcessingPipeline:
...
+                continue
+            if "analysis" not in token:
+                continue
+            morph_token = MorphologicalToken(original_word=token['text'])
+            morph_token.normalized_form = token['analysis'][0]['lex']
```
Comment (demid5111) [inline pipeline.py:136]: too many assumptions

File: pipeline.py [hunk 31]
```diff
@@ -72,31 +91,100 @@ class TextProcessingPipeline:
...
+            morph_token.tags_mystem = token['analysis'][0]['gr']
+            morph_token.tags_pymorphy = morphs.parse(token['text'])[0].tag
```
Comment (demid5111) [inline pipeline.py:138]: no checks (!)

File: pipeline.py [hunk 32]
```diff
@@ -72,31 +91,100 @@ class TextProcessingPipeline:
...
+
+    data = list(path_to_validate.glob('*'))
+
+    if not data:
+        raise EmptyDirectoryError
+
+    number_of_texts = 0
+    number_of_metas = 0
+
+    pattern = re.compile(r'[0-9]+')
```
Comment (demid5111) [inline pipeline.py]: .

File: pipeline.py [hunk 33]
```diff
@@ -72,31 +91,100 @@ class TextProcessingPipeline:
...
+    for file in sorted(list(path_to_validate.glob('*')), key=lambda x: int(pattern.search(x.name).group(0))):
+        if "_raw.txt" in file.name:
```
Comment (demid5111) [inline pipeline.py]: nested

File: pipeline.py [hunk 34]
```diff
@@ -72,31 +91,100 @@ class TextProcessingPipeline:
...
+    data = list(path_to_validate.glob('*'))
+
+    if not data:
+        raise EmptyDirectoryError
+
+    number_of_texts = 0
+    number_of_metas = 0
+
+    pattern = re.compile(r'[0-9]+')
+    for file in sorted(list(path_to_validate.glob('*')), key=lambda x: int(pattern.search(x.name).group(0))):
```
Comment (demid5111) [inline pipeline.py]: too reckless

File: pipeline.py [hunk 35]
```diff
@@ -72,31 +91,100 @@ class TextProcessingPipeline:
...
+        if "_raw.txt" in file.name:
+            with open(file, 'r', encoding='utf-8') as text_file:
```
Comment (demid5111) [inline pipeline.py]: should be a better way to understand whether a file is empty. ask colleagues if you get stuck

File: pipeline.py [hunk 36]
```diff
@@ -72,31 +91,100 @@ class TextProcessingPipeline:
...
+                text = text_file.read()
+                if not text:
+                    raise InconsistentDatasetError
+            file_id = int(pattern.search(file.name).group(0))
+            if file_id == 0 or file_id - number_of_texts > 1 or \
```
Comment (demid5111) [inline pipeline.py]: check that ids start from 1

File: pos_frequency_pipeline.py [hunk 37]
```diff
@@ -13,15 +19,38 @@ class EmptyFileError(Exception):
 
 class POSFrequencyPipeline:
     def __init__(self, corpus_manager: CorpusManager):
-        pass
+        self.corpus_manager = corpus_manager
 
     def run(self):
-        pass
+        articles = self.corpus_manager.get_articles()
```
Comment (demid5111) [inline pos_frequency_pipeline.py]: .

**Review by demid5111 — CHANGES_REQUESTED** [review]

File: pipeline.py [hunk 38]
```diff
@@ -72,31 +91,94 @@ class TextProcessingPipeline:
...
+    if not path_to_validate.is_dir():
+        raise NotADirectoryError
+
+    data = list(path_to_validate.glob('*'))
+
+    if not data:
+        raise EmptyDirectoryError
+
+    pattern = re.compile(r'\d+')
+    raw_texts = sorted(list(path_to_validate.glob('*_raw.txt')), key=lambda x: int(pattern.search(x.name).group(0)))
```
Comment (demid5111) [inline pipeline.py]: numbers can be absent (!)

**Review by demid5111 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**